examples/microphone_nextjs/utils/recorder.ts

```diff
@@ -0,0 +1,174 @@
+  constructor(dataHandlerCallback: (data: Blob) => void) {
+    this.dataHandlerCallback = dataHandlerCallback;
+  }
+
+  async startRecording(deviceId: string) {
+    const AudioContext = globalThis.window?.AudioContext;
+    this.audioContext = new AudioContext({ sampleRate: SAMPLE_RATE_48K });
+
+    // We first check mic permissions in case they are explicitly denied
+    if ((await getPermissions()) === "denied") {
+      throw new Error("Microphone permission denied.");
+    }
+
+    // Here we set the sample rate and the deviceId that the user has selected
+    // If deviceId isn't set, then we get a default device (which is expected behaviour)
+    const audio: MediaTrackConstraintSet = {
+      sampleRate: SAMPLE_RATE_48K,
+      deviceId,
+    };
+
+    // Now we open and store the stream
+    this.stream = await navigator.mediaDevices.getUserMedia({ audio });
+
+    // Instantiate the MediaRecorder instance
+    this.recorder = new MediaRecorder(this.stream);
+
+    // This is the event listening function that gets called when data is available
+    this.recorder.ondataavailable = (event: BlobEvent) => {
+      this.dataHandlerCallback?.(event.data);
+    };
+
+    // Start recording from the device
+    // The number passed in indicates how frequently in milliseconds ondataavailable will be called
+    this.recorder.start(500);
+
+    // return the sample rate
+    return { sampleRate: this.audioContext.sampleRate };
+  }
+
+  // stopRecording is called when the session ends
+  // It shuts down the stream and recorder and sets all properties to null
+  async stopRecording() {
+    this.mediaStreamSource?.disconnect();
+    this.recorder?.stop();
+    this.stopStream();
+    this.resetRecordingProperties();
+  }
+
+  // stopStream stops all tracks in the stream
+  private stopStream() {
+    this.stream?.getTracks().forEach((track) => track.stop()); //stop each one
+  }
+
+  // resetRecordingProperties makes sure we have a clean slate for the next session startup
+  private resetRecordingProperties() {
+    this.stream = null;
+    this.mediaStreamSource = null;
+  }
+}
+
+// useAudioDevices is the hook we'll use to get a list of all the user's microphones
+// requested provides us with a convenient way to control whether we ask for permissions on load
+// or on a user action e.g. clicking the select menu drop down
+export function useAudioDevices(
+  requested = false
+): [MediaDeviceInfo[], boolean] {
+  const [devices, setDevices] = useState<MediaDeviceInfo[]>([]);
+  const [denied, setIsDenied] = useState<boolean>(false);
+
+  // Create the async function to set devices
+  const getDevices = async () => {
+    try {
+      const devs = await getAudioInputs();
+      setDevices(devs);
+    } catch (err) {
+      setIsDenied(true);
+    }
+  };
+
+  // Call the async method in useEffect
+  useEffect(() => {
+    if (requested) getDevices();
+  }, [requested]);
+
+  return [devices, denied];
+}
```

Comment:
A lot of demo code that accesses getUserMedia will use useEffect, but I think the recommended way to do this with modern React is to use useExternalStore, which also brings some advantages.

- https://react.dev/learn/you-might-not-need-an-effect
- https://react.dev/reference/react/useSyncExternalStore#subscribing-to-a-browser-api
- https://github.com/penx/use-effect-cheatsheet

I have something working locally that will do this, I'll tidy up and raise a PR in to your branch.

Needs some testing, but hopefully my incoming PR means:

- the device list can be auto updated when users plug in or remove devices
- the device list is available on initial render rather than waiting for the user to click the drop down
- robust when multiple calls in parallel
- works with double rendering of hooks in dev mode


The main change would be replacing useAudioDevices with this:

```ts

// this is a Class so that we can use EventTarget, but a singleton, as in the browser the active devices are external to React and can be managed app-wide.
class AudioDevices extends EventTarget {
  public devices: MediaDeviceInfo[] = [];
  public denied = false;
  private busy = false;

  constructor() {
    super();
    if (typeof window !== "undefined") {
      this.getDevices();
      navigator.mediaDevices.ondevicechange = () => {
        this.getDevices();
      };
    }
  }

  public getDevices = async () => {
    if (!this.busy) {
      this.busy = true;
      try {
        const devs = await getAudioInputs();
        this.devices = devs;
        this.denied = false;
      } catch (err) {
        this.denied = true;
      } finally {
        this.busy = false;
        this.dispatchEvent(new Event("change"));
      }
    }
  };
}
const audioDevices = new AudioDevices();

function subscribeAudioDevices(callback) {
  audioDevices.addEventListener("change", callback);
  return () => {
    audioDevices.removeEventListener("change", callback);
  };
}
export function useAudioDevices() {
  return useSyncExternalStore(
    subscribeAudioDevices,
    () => audioDevices.devices,
    () => audioDevices.devices
  );
}
export function useAudioDenied() {
  return useSyncExternalStore(
    subscribeAudioDevices,
    () => audioDevices.denied,
    () => audioDevices.denied
  );
}
```

Reply:
Okay that looks good, thanks for the tip! One thing to note is that not accessing the media straight away is a conscious choice because I was thinking of giving the user the option of using the default audio device without having to worry about selecting an input. I was thinking the user experience was better when permissions were only prompted after direct user input rather than on page load. Not sure your thoughts on this?

Reply:
Yes that makes sense.

I was finding it a little janky clicking on the select and having it reload its options.

Perhaps, if a user has already granted access, we could get the device names via enumerateDevices on page load, but wait for a user interaction before we try to get them via a prompt.